### PR TITLE
Fix false positive for `CanBeNonNullable` rule

### PR DESCRIPTION
**Pull request by atulgpt**
Added failing TCs
Fixes https://github.com/detekt/detekt/issues/5629

### CODE REVIEW

**Comment by BraisGabin** [comment]
Are these tests related with #5629? The cases are a bit difficult to follow. And I'm not sure if all of them are related with the same issue.

Would it be possible to split them in different PRs? Or all of them target the same flaw?

**Comment by atulgpt** [comment]
Hi, @BraisGabin not all follow the same flaw. Some can be grouped and split into little PRs. Is there any provision to add disabled test in the PR? Or should I completely remove them from here and add in the other PRs?

**Comment by BraisGabin** [comment]
I would prefer to split them up.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt [hunk 1]
```diff
@@ -720,6 +793,193 @@ class CanBeNonNullableSpec(val env: KotlinCoreEnvironment) {
...
+                }
+
+                @Test
+                fun `does report when parameter is not checked for nullability with inc operator`() {
+                    val code = """
+                        class A {
+                            operator fun not() = this
+                        }
+                        
+
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt:876]:
```suggestion
```

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt [hunk 2]
```diff
@@ -720,6 +793,193 @@ class CanBeNonNullableSpec(val env: KotlinCoreEnvironment) {
...
+                                (b != null) && !b != a -> {
+                                    println("doesn't handle null")
+                                }
+                                (c > 10 || (b != null && !b != a)) -> {
+                                    println("doesn't handle null")
+                                }
+                            }
+                        }
+                    """.trimIndent()
+                    assertThat(subject.compileAndLintWithContext(env, code)).hasSize(1)
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt:892]: I dont't get what this test does exactly.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt [hunk 3]
```diff
@@ -774,6 +1034,36 @@ class CanBeNonNullableSpec(val env: KotlinCoreEnvironment) {
                     """.trimIndent()
                     assertThat(subject.compileAndLintWithContext(env, code)).isEmpty()
                 }
+
+                // Failing
+                @Test
+                fun `does not report when else handles null case`() {
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt:1040]: Isn't this test exactly the same as `does not report when else expression variable is smart casted to null`?

**Review by BraisGabin — COMMENTED** [review]
Hi @atulgpt which are the next steps for this PR? Is there somerhing blocking you here? Can I help you some how? I left some comments on the tests but they seem fine.

I must say that a function that only does something when a parameter is null seems like a function that should be flagged too. But probably not by this rule but by other one. But this is offtopic for this PR.

**Comment by atulgpt** [comment]
> Hi @atulgpt which are the next steps for this PR? Is there somerhing blocking you here? Can I help you some how? I left some comments on the tests but they seem fine.
> 
> I must say that a function that only does something when a parameter is null seems like a function that should be flagged too. But probably not by this rule but by other one. But this is offtopic for this PR.

Hi @BraisGabin good vibes of Bali is stopping me from completing this PR 😂. I will complete back after I come back 🙂

**Comment by BraisGabin** [comment]
> Hi @BraisGabin good vibes of Bali is stopping me from completing this PR 😂. I will complete back after I come back 🙂

Enjoy!!!

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt (hunk 2 above)
Comment (atulgpt) [reply detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt:891]: The name of the TC was wrong. Actually, it should report as all the branches of `when` handles the non-null case. Here we are checking different types of boolean conditions. I have renamed the TC to `does report when parameter is not checked for nullability in boolean conditions`

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullable.kt [hunk 4]
```diff
@@ -231,13 +233,22 @@ class CanBeNonNullable(config: Config = Config.empty) : Rule(config) {
         }
 
         override fun visitWhenExpression(expression: KtWhenExpression) {
-            val subjectDescriptor = expression.subjectExpression
-                ?.let { it as? KtNameReferenceExpression }
-                ?.getResolvedCall(bindingContext)
-                ?.resultingDescriptor
+            val nullCheckedDescriptor =
```
Comment (atulgpt) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullable.kt:236]: Hi, @BraisGabin this fixes the current issue but this solution is not a robust solution and will give a false negative when Kotlin `when` subject expression contains complex logic.

Consider the below example
```
fun foo(a: Int?, b: Int?) {
    when (a?.also {
        println(b)
    }?.plus(4)) {
        null -> println("We have handled a is null as well")
    }
}
```

for the above example, the rule should ideally give one warning but it gives no warning.

Actually, I am kind of stuck as to how I can determine if `b`(or any variable present in `when` expression) contributes to the nullability of the whole `when` subject expression or not. In this example, `b` is not contributing to the nullability of the below `when` subject expression
```
a?.also {
        println(b)
    }?.plus(4)
```.

But this code handles the issue mentioned in the PR. Either we can completely handle all the case or fix the rule case by case

Comment (BraisGabin) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullable.kt:236]: This rule had some false positives already. Is this one of those or is this a new one?

If this improves the rule and doesn't introduce any regression I would merge it. It is not perfect? OK, we can work on that in next PRs but our users can benefit from this improvements meanwhile.

Also, complex expressions inside when's is not a very common case so we shouldn't expand far too much time there.

Comment (atulgpt) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullable.kt:236]: `This rule had some false positives already. Is this one of those or is this a new one?` These are existing ones but initially they were false negative with more extent now some cases has been fix and some has been converted to false postive

Comment (BraisGabin) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullable.kt:236]: 🤔 The false positive are a problem... We shouldn't add of those. Any way to get rid of them?

Comment (atulgpt) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullable.kt:236]: Edited my comment. Please see the edited comment

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullable.kt [hunk 5]
```diff
@@ -231,13 +233,22 @@ class CanBeNonNullable(config: Config = Config.empty) : Rule(config) {
...
+                        val callDescriptor =
+                            it.getResolvedCall(bindingContext)
+                        val valueParameterDescriptor =
+                            callDescriptor?.resultingDescriptor as? ValueParameterDescriptor
+                        if (valueParameterDescriptor != null && callDescriptor.getReturnType().isNullable()) {
+                            valueParameterDescriptor
+                        } else {
+                            null
+                        }
+                    }
```
Comment (BraisGabin) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullable.kt:248]:
```suggestion
            val nullCheckedDescriptor = expression.subjectExpression?
                .collectDescendantsOfType<KtNameReferenceExpression>()
                .orEmpty()
                .mapNotNull { it.getResolvedCall(bindingContext) }
                .filter { callDescriptor -> callDescriptor.getReturnType().isNullable() }
                .mapNotNull { callDescriptor -> callDescriptor.resultingDescriptor as? ValueParameterDescriptor }
```

What do you think? Is it easier to read?

**Review by BraisGabin — APPROVED** [review]
I left some nit comments but LGTM

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullable.kt (hunk 4 above)
Comment (BraisGabin) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullable.kt:236]: Then we can merge it. Later we could improve it even more.

**Comment by BraisGabin** [comment]
It seems that my code didn't compile 😅. The Github's suggestions feature is great but working without an IDE is difficult.

**Comment by atulgpt** [comment]
> It seems that my code didn't compile 😅. The Github's suggestions feature is great but working without an IDE is difficult.

Haha true 😅 I have fixed it to make it compliable. Now CI should pass

**Review by cortinico — APPROVED** [review]

**Pull request merged**